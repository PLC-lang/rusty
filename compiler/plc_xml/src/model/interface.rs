use std::borrow::Borrow;

use quick_xml::events::{BytesStart, Event};

use crate::error::Error;
use crate::reader::Reader;
use crate::xml_parser::Parseable;

use super::pou::PouType;

#[derive(Debug)]
pub(crate) struct Interface {
    //TODO: return type, variables, etc..
    // we only care about the addData field for now, as it contains our text-declaration

    // the addData field is be application-specific and can be implemented/extended as needed
    add_data: Option<Data>,
}

// XXX: this implementation is very specific to our own format. we might want to make it generic/provide an interface via traits
// so it can be easily modified/extended for other purposes
impl Interface {
    pub fn new(content: &str) -> Self {
        Interface { add_data: Some(Data::new_implementation(content)) }
    }

    pub fn get_data_content(&self) -> Option<&str> {
        let Some(ref data) = self.add_data else { return None };

        Some(&data.content)
    }

    // We have to append a END_... to the declaration, as it is missing in our text declaration
    pub fn append_end_keyword(self, pou_type: &PouType) -> Self {
        let Some(old_data) = self.add_data else {
            // if we have no content, we have nothing to append to. return as is
            return self;
        };

        Interface {
            add_data: Some(Data::new_implementation(&format!("{}\nEND_{}", old_data.content, pou_type))),
        }
    }
}

impl Parseable for Interface {
    fn visit(reader: &mut Reader, _tag: Option<BytesStart>) -> Result<Self, Error> {
        let mut interface = Interface { add_data: None };
        loop {
            match reader.read_event().map_err(Error::ReadEvent)? {
                Event::End(tag) if tag.name().as_ref() == b"interface" => break,
                Event::Text(text) => {
                    interface.add_data =
                        Some(Data::new_implementation(text.unescape().map_err(Error::ReadEvent)?.borrow()))
                }
                _ => {}
            }
        }
        Ok(interface)
    }
}

/// Application specific data
#[derive(Debug)]
struct Data {
    content: String,
    handle: HandleUnknown,
}

impl Data {
    fn new_implementation(content: &str) -> Self {
        Data { content: content.to_string(), handle: HandleUnknown::Implementation }
    }
}

/// Recommended processor-handling for unknown data elements
#[derive(Debug)]
enum HandleUnknown {
    Preserve,
    Discard,
    Implementation,
}
