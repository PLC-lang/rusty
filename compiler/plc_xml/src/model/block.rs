use std::borrow::Cow;

use quick_xml::events::{BytesStart, Event};
use rustc_hash::FxHashMap;

use crate::{
    error::Error,
    extensions::GetOrErr,
    reader::Reader,
    xml_parser::{get_attributes, Parseable},
};

use super::variables::BlockVariable;

#[derive(Debug, PartialEq, Eq, Hash)]
pub(crate) struct Block<'xml> {
    pub local_id: usize,
    pub type_name: Cow<'xml, str>,
    pub instance_name: Option<Cow<'xml, str>>,
    pub execution_order_id: Option<usize>,
    pub variables: Vec<BlockVariable>,
}

impl<'xml> Block<'xml> {
    pub fn new(mut hm: FxHashMap<String, String>, variables: Vec<BlockVariable>) -> Result<Self, Error> {
        Ok(Self {
            local_id: hm.get_or_err("localId").map(|it| it.parse())??,
            type_name: Cow::from(hm.get_or_err("typeName")?),
            instance_name: hm.remove("instanceName").map(Cow::from),
            execution_order_id: hm.get("executionOrderId").map(|it| it.parse()).transpose()?,
            variables,
        })
    }
}

impl<'xml> Parseable for Block<'xml> {
    fn visit(reader: &mut Reader, tag: Option<BytesStart>) -> Result<Self, Error> {
        let Some(tag) = tag else { unreachable!() };
        let attributes = get_attributes(tag.attributes())?;
        let mut variables = Vec::new();

        loop {
            match reader.read_event().map_err(Error::ReadEvent)? {
                Event::Start(tag) => match tag.name().as_ref() {
                    b"inputVariables" | b"outputVariables" | b"inOutVariables" => {
                        let new_vars: Vec<BlockVariable> = Parseable::visit(reader, Some(tag))?;
                        variables.extend(new_vars);
                    }
                    _ => {}
                },

                Event::End(tag) if tag.name().as_ref() == b"block" => {
                    break;
                }

                Event::Eof => return Err(Error::UnexpectedEndOfFile(vec![b"block"])),
                _ => {}
            }
        }

        Block::new(attributes, variables)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::{
        model::block::Block,
        reader::{get_start_tag, Reader},
        serializer::{SBlock, SVariable},
        xml_parser::Parseable,
    };

    #[test]
    fn add_block() {
        let content = SBlock::init("ADD", 1, 0)
            .with_input(vec![
                &SVariable::new().with_name("a").connect(1),
                &SVariable::new().with_name("b").connect(2),
            ])
            .with_output(vec![&SVariable::new().with_name("c")])
            .serialize();

        let mut reader = Reader::new(&content);
        let tag = get_start_tag(reader.read_event().unwrap());
        assert_debug_snapshot!(Block::visit(&mut reader, tag).unwrap());
    }
}
