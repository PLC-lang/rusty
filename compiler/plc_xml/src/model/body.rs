use quick_xml::events::{BytesStart, Event};

use super::fbd::FunctionBlockDiagram;
use crate::{error::Error, reader::Reader, xml_parser::Parseable};

#[derive(Debug, Default)]
pub(crate) struct Body<'xml> {
    pub function_block_diagram: FunctionBlockDiagram<'xml>,
}

impl<'xml> Body<'xml> {
    fn new(fbd: FunctionBlockDiagram<'xml>) -> Result<Self, Error> {
        Ok(Self { function_block_diagram: fbd })
    }

    fn empty() -> Result<Self, Error> {
        Ok(Self { function_block_diagram: FunctionBlockDiagram::default() })
    }
}

impl<'xml> Parseable for Body<'xml> {
    fn visit(reader: &mut Reader, _tag: Option<BytesStart>) -> Result<Self, Error> {
        let mut body = Body::default();
        loop {
            match reader.read_event().map_err(Error::ReadEvent)? {
                Event::Start(tag) if tag.name().as_ref() == b"FBD" => {
                    body.function_block_diagram = FunctionBlockDiagram::visit(reader, Some(tag))?
                }
                Event::End(tag) if tag.name().as_ref() == b"body" => break,
                Event::Eof => return Err(Error::UnexpectedEndOfFile(vec![b"body"])),
                _ => {}
            }
        }

        Ok(body)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::{
        model::body::Body,
        reader::{get_start_tag, Reader},
        serializer::{XBlock, XBody, XFbd, XInOutVariables, XInputVariables, XOutputVariables, XVariable},
        xml_parser::Parseable,
    };

    #[test]
    fn empty() {
        let content = XBody::new().with_fbd(XFbd::new().close()).serialize();

        let mut reader = Reader::new(&content);
        assert_debug_snapshot!(Body::visit(&mut reader, None).unwrap());
    }

    #[test]
    fn fbd_with_add_block() {
        let content = XBody::new()
            .with_fbd(
                XFbd::new().with_block(
                    XBlock::init("1", "ADD", "0")
                        .with_input_variables(
                            XInputVariables::new()
                                .with_variable(
                                    XVariable::init("a", false).with_connection_in_initialized("1"),
                                )
                                .with_variable(
                                    XVariable::init("b", false).with_connection_in_initialized("2"),
                                ),
                        )
                        .with_inout_variables(XInOutVariables::new().close())
                        .with_output_variables(
                            XOutputVariables::new()
                                .with_variable(XVariable::init("c", false).with_connection_out_initialized()),
                        ),
                ),
            )
            .serialize();

        let mut reader = Reader::new(&content);
        let tag = get_start_tag(reader.read_event().unwrap());
        assert_debug_snapshot!(Body::visit(&mut reader, tag).unwrap());
    }
}
