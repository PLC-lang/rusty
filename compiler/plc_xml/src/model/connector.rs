use std::borrow::Cow;

use quick_xml::events::Event;
use rustc_hash::FxHashMap;

use crate::{
    error::Error,
    extensions::{GetOrErr, TryToString},
    reader::Reader,
    xml_parser::{get_attributes, Parseable},
};

#[derive(Debug, PartialEq, Eq, Hash)]
pub(crate) struct Connector<'xml> {
    pub kind: ConnectorKind,
    pub name: Cow<'xml, str>,
    pub local_id: usize,
    pub ref_local_id: Option<usize>,
    pub formal_parameter: Option<Cow<'xml, str>>,
}

impl<'xml> Connector<'xml> {
    pub fn new(mut hm: FxHashMap<String, String>, kind: ConnectorKind) -> Result<Self, Error> {
        Ok(Self {
            kind,
            name: Cow::from(hm.get_or_err("name")?),
            local_id: hm.get_or_err("localId").map(|it| it.parse())??,
            ref_local_id: hm.get("refLocalId").map(|it| it.parse()).transpose()?,
            formal_parameter: hm.remove("formalParameter").map(Cow::from),
        })
    }
}

#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
pub(crate) enum ConnectorKind {
    Source,
    Sink,
}

impl<'xml> Parseable for Connector<'xml> {
    fn visit(reader: &mut Reader, tag: Option<quick_xml::events::BytesStart>) -> Result<Self, Error> {
        let Some(tag) = tag else { unreachable!() };

        let kind = match tag.name().as_ref() {
            b"connector" => ConnectorKind::Source,
            b"continuation" => ConnectorKind::Sink,
            _ => return Err(Error::UnexpectedElement(tag.name().try_to_string()?)),
        };

        let mut attributes = get_attributes(tag.attributes())?;
        loop {
            match reader.read_event().map_err(Error::ReadEvent)? {
                Event::Start(tag) | Event::Empty(tag) if tag.name().as_ref() == b"connection" => {
                    attributes.extend(get_attributes(tag.attributes())?)
                }
                Event::End(tag) if matches!(tag.name().as_ref(), b"connector" | b"continuation") => break,
                _ => {}
            }
        }
        Connector::new(attributes, kind)
    }
}
