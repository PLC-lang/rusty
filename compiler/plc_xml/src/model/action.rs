use std::borrow::Cow;

use quick_xml::events::Event;

use crate::{
    error::Error,
    reader::Reader,
    xml_parser::{get_attributes, Parseable},
};

use super::body::Body;

#[derive(Debug, Default)]
pub(crate) struct Action<'xml> {
    pub name: Cow<'xml, str>,
    pub type_name: Cow<'xml, str>,
    pub body: Body<'xml>,
}

impl Action<'_> {
    pub fn new(name: &str) -> Self {
        Action { name: name.to_string().into(), ..Default::default() }
    }
}

impl Parseable for Vec<Action<'_>> {
    fn visit(
        reader: &mut Reader,
        _tag: Option<quick_xml::events::BytesStart>,
    ) -> Result<Self, crate::error::Error> {
        let mut res = vec![];
        loop {
            match reader.read_event()? {
                Event::Start(tag) if tag.name().as_ref() == b"action" => {
                    res.push(Action::visit(reader, Some(tag))?);
                }
                Event::End(tag) if tag.name().as_ref() == b"actions" => break,
                Event::Eof => return Err(Error::UnexpectedEndOfFile(vec![b"actions"])),
                _ => {}
            }
        }

        Ok(res)
    }
}

impl Parseable for Action<'_> {
    fn visit(
        reader: &mut Reader,
        tag: Option<quick_xml::events::BytesStart>,
    ) -> Result<Self, crate::error::Error> {
        let Some(tag) = tag else { unreachable!() };

        let attributes = get_attributes(tag.attributes())?;
        let Some(name) = attributes.get("name") else { todo!() };
        let mut action = Action::new(name);
        loop {
            match reader.read_event()? {
                Event::Start(tag) if tag.name().as_ref() == b"body" => {
                    action.body = Body::visit(reader, Some(tag))?;
                }
                Event::End(tag) if tag.name().as_ref() == b"action" => break,
                Event::Eof => return Err(Error::UnexpectedEndOfFile(vec![b"actions"])),
                _ => {}
            }
        }

        Ok(action)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::xml_parser;

    #[test]
    fn list_of_actions_parsed_to_model() {
        let src = r###"
<?xml version="1.0" encoding="UTF-8"?>
<pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
    <interface>
        <localVars/>
        <addData>
            <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                <textDeclaration>
                    <content>
PROGRAM program_0
VAR
    a : DINT;
END_VAR
                    </content>
                </textDeclaration>
            </data>
        </addData>
    </interface>
    <actions>
        <action name="newAction">
            <body>
                <FBD>
                    <block localId="1" width="80" height="60" typeName="foo" executionOrderId="0">
                        <position x="420" y="110"/>
                        <inputVariables>
                            <variable formalParameter="IN1" negated="false">
                                <connectionPointIn>
                                    <relPosition x="0" y="30"/>
                                </connectionPointIn>
                            </variable>
                            <variable formalParameter="IN2" negated="false">
                                <connectionPointIn>
                                    <relPosition x="0" y="50"/>
                                </connectionPointIn>
                            </variable>
                        </inputVariables>
                        <inOutVariables/>
                        <outputVariables>
                            <variable formalParameter="OUT" negated="false">
                                <connectionPointOut>
                                    <relPosition x="80" y="30"/>
                                </connectionPointOut>
                            </variable>
                        </outputVariables>
                    </block>
                </FBD>
            </body>
        </action>
        <action name="newAction2">
            <body>
                <FBD>
                    <inOutVariable localId="1" height="20" width="80" negatedIn="false" storageIn="none" negatedOut="false">
                        <position x="200" y="70"/>
                        <connectionPointIn>
                            <relPosition x="0" y="10"/>
                        </connectionPointIn>
                        <connectionPointOut>
                            <relPosition x="80" y="10"/>
                        </connectionPointOut>
                        <expression>a</expression>
                    </inOutVariable>
                </FBD>
            </body>
        </action>
    </actions>
    <body>
        <FBD/>
    </body>
</pou>
        "###;

        assert_debug_snapshot!(xml_parser::visit(src).unwrap())
    }
}
