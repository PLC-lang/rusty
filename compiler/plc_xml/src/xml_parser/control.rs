use ast::ast::{AstFactory, AstNode};
use plc_diagnostics::diagnostics::Diagnostic;

use crate::model::{
    control::{Control, ControlKind},
    fbd::{Node, NodeIndex},
};

use super::ParseSession;

impl<'xml> Control<'xml> {
    pub(crate) fn transform(&self, session: &ParseSession, index: &NodeIndex) -> Result<AstNode, Diagnostic> {
        match self.kind {
            ControlKind::Jump => transform_jump(self, session, index),
            ControlKind::Label => transform_label(self, session),
            ControlKind::Return => transform_return(self, session, index),
        }
    }
}

fn get_connection(
    control: &Control,
    session: &ParseSession,
    index: &NodeIndex,
) -> Result<AstNode, Diagnostic> {
    let Some(ref_local_id) = control.ref_local_id else {
        let location =
            session.range_factory.create_block_location(control.local_id, control.execution_order_id);
        return Err(Diagnostic::empty_control_statement(location));
    };

    let Some(node) = index.get(&ref_local_id) else {
        let location = session.range_factory.create_block_location(ref_local_id, None);
        return Err(Diagnostic::undefined_node(control.local_id, ref_local_id, location));
    };

    match node {
        Node::FunctionBlockVariable(variable) => Ok(variable.transform(session)),
        Node::Block(block) => Ok(block.transform(session, index)),

        _ => {
            let location_control =
                session.range_factory.create_block_location(control.local_id, control.execution_order_id);
            let location_other =
                session.range_factory.create_block_location(ref_local_id, node.get_exec_id());

            Err(Diagnostic::unexpected_nodes(location_control.span(&location_other)))
        }
    }
}

fn transform_jump(
    control: &Control,
    session: &ParseSession,
    index: &NodeIndex,
) -> Result<AstNode, Diagnostic> {
    let location = session.range_factory.create_block_location(control.local_id, control.execution_order_id);
    let condition = get_connection(control, session, index)?;
    let condition = if control.negated { condition.negate(session.id_provider.clone()) } else { condition };

    let target = control
        .name
        .as_ref()
        .filter(|it| !it.is_empty())
        .map(|it| session.parse_expression(it.as_ref(), control.local_id, control.execution_order_id))
        .ok_or_else(|| Diagnostic::unnamed_control(location.clone()))?;

    Ok(AstFactory::create_jump_statement(Box::new(condition), Box::new(target), location, session.next_id()))
}

fn transform_label(control: &Control, session: &ParseSession) -> Result<AstNode, Diagnostic> {
    let location = session.range_factory.create_block_location(control.local_id, control.execution_order_id);

    control
        .name
        .as_ref()
        .filter(|it| !it.is_empty())
        .map(|it| AstFactory::create_label_statement(it.to_string(), location.clone(), session.next_id()))
        .ok_or_else(|| Diagnostic::unnamed_control(location))
}

fn transform_return(
    control: &Control,
    session: &ParseSession,
    index: &NodeIndex,
) -> Result<AstNode, Diagnostic> {
    let condition = get_connection(control, session, index)?;
    let condition = if control.negated { condition.negate(session.id_provider.clone()) } else { condition };

    Ok(AstFactory::create_return_statement(
        Some(condition),
        session.create_block_location(control.local_id, control.execution_order_id),
        session.next_id(),
    ))
}

#[cfg(test)]
mod tests {
    use ast::provider::IdProvider;
    use insta::assert_debug_snapshot;

    use crate::{
        model::control::Control,
        reader::{get_start_tag, Reader},
        serializer::{XAddData, XConnectionPointIn, XJump, XLabel, XReturn},
        xml_parser::{self, Parseable},
    };

    #[test]
    fn simple_return() {
        let content = XReturn::new()
            .with_local_id("1")
            .with_execution_order_id("2")
            .with_connection_point_in(XConnectionPointIn::with_ref("3"))
            .with_add_data(XAddData::negated(false))
            .serialize();

        let reader = &mut Reader::new(&content);
        let tag = get_start_tag(reader.read_event().unwrap());
        assert_debug_snapshot!(Control::visit(reader, tag).unwrap());
    }

    #[test]
    fn simple_negated_return() {
        let content = XReturn::new()
            .with_local_id("1")
            .with_execution_order_id("2")
            .with_connection_point_in(XConnectionPointIn::with_ref("3"))
            .with_add_data(XAddData::negated(true))
            .serialize();

        let reader = &mut Reader::new(&content);
        let tag = get_start_tag(reader.read_event().unwrap());
        assert_debug_snapshot!(Control::visit(reader, tag).unwrap());
    }

    #[test]
    fn jump_to_label() {
        let content = r###"
            <?xml version="1.0" encoding="UTF-8"?>
            <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
                <interface>
                    <localVars/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <textDeclaration>
                                <content>
                                    PROGRAM program_0
                                    VAR&#xd;
                                        x: BOOL := 0;&#xd;
                                    END_VAR
                                </content>
                            </textDeclaration>
                        </data>
                    </addData>
                </interface>
                <body>
                    <FBD>
                        <inVariable localId="0" height="20" width="80" negated="false">
                            <position x="160" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>x</expression>
                        </inVariable>
                        <label localId="1" height="20" width="80" label="lbl" executionOrderId="0">
                            <position x="570" y="50"/>
                        </label>
                        <jump localId="2" height="20" width="80" label="lbl" executionOrderId="1">
                            <position x="320" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="0"/>
                            </connectionPointIn>
                            <addData>
                                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                                    <negated value="false"/>
                                </data>
                            </addData>
                        </jump>
                        <outVariable localId="3" height="20" width="80" executionOrderId="2" negated="false" storage="none">
                            <position x="790" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="4"/>
                            </connectionPointIn>
                            <expression>x</expression>
                        </outVariable>
                        <inVariable localId="4" height="20" width="80" negated="false">
                            <position x="640" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>FALSE</expression>
                        </inVariable>
                    </FBD>
                </body>
            </pou>
        "###;

        assert_debug_snapshot!(xml_parser::visit(content));
    }

    #[test]
    fn unconnected_jump_to_label() {
        let content = XJump::new().with_local_id("1").with_execution_order_id("2").serialize();

        let mut reader = Reader::new(&content);
        let tag = get_start_tag(reader.read_event().unwrap());
        assert_debug_snapshot!(Control::visit(&mut reader, tag).unwrap());
    }

    #[test]
    fn negated_jump() {
        let content = r###"
            <?xml version="1.0" encoding="UTF-8"?>
            <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
                <interface>
                    <localVars/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <textDeclaration>
                                <content>
                                    PROGRAM program_0
                                    VAR&#xd;
                                        x: BOOL := 0;&#xd;
                                    END_VAR
                                </content>
                            </textDeclaration>
                        </data>
                    </addData>
                </interface>
                <body>
                    <FBD>
                        <inVariable localId="0" height="20" width="80" negated="false">
                            <position x="160" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>x</expression>
                        </inVariable>
                        <label localId="1" height="20" width="80" label="lbl" executionOrderId="0">
                            <position x="570" y="50"/>
                        </label>
                        <jump localId="2" height="20" width="80" label="lbl" executionOrderId="1">
                            <position x="320" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="0"/>
                            </connectionPointIn>
                            <addData>
                                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                                    <negated value="true"/>
                                </data>
                            </addData>
                        </jump>
                        <outVariable localId="3" height="20" width="80" executionOrderId="2" negated="false" storage="none">
                            <position x="790" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="4"/>
                            </connectionPointIn>
                            <expression>x</expression>
                        </outVariable>
                        <inVariable localId="4" height="20" width="80" negated="false">
                            <position x="640" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>FALSE</expression>
                        </inVariable>
                    </FBD>
                </body>
            </pou>
        "###;

        assert_debug_snapshot!(xml_parser::visit(content));
    }

    #[test]
    fn label_parsed() {
        let content = XLabel::new().with_local_id("1").with_execution_order_id("2").serialize();

        let mut reader = Reader::new(&content);
        let tag = get_start_tag(reader.read_event().unwrap());
        assert_debug_snapshot!(Control::visit(&mut reader, tag).unwrap());
    }

    #[test]
    fn jump_and_label_converted_to_ast() {
        let content = r###"
            <?xml version="1.0" encoding="UTF-8"?>
            <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
                <interface>
                    <localVars/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <textDeclaration>
                                <content>
                                    PROGRAM program_0
                                    VAR
                                        x: BOOL := 0;
                                    END_VAR
                                </content>
                            </textDeclaration>
                        </data>
                    </addData>
                </interface>
                <body>
                    <FBD>
                        <inVariable localId="0" height="20" width="80" negated="false">
                            <position x="160" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>x</expression>
                        </inVariable>
                        <label localId="1" height="20" width="80" label="lbl" executionOrderId="0">
                            <position x="570" y="50"/>
                        </label>
                        <jump localId="2" height="20" width="80" label="lbl" executionOrderId="1">
                            <position x="320" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="0"/>
                            </connectionPointIn>
                            <addData>
                                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                                    <negated value="false"/>
                                </data>
                            </addData>
                        </jump>
                        <outVariable localId="3" height="20" width="80" executionOrderId="2" negated="false" storage="none">
                            <position x="790" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="4"/>
                            </connectionPointIn>
                            <expression>x</expression>
                        </outVariable>
                        <inVariable localId="4" height="20" width="80" negated="false">
                            <position x="640" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>FALSE</expression>
                        </inVariable>
                    </FBD>
                </body>
            </pou>
        "###.to_string();

        let (ast, diagnostics) =
            xml_parser::parse(&content.into(), ast::ast::LinkageType::Internal, IdProvider::default());
        assert_debug_snapshot!(ast);
        assert!(diagnostics.is_empty());
    }

    #[test]
    fn unconnected_jump_generated_as_empty_statement() {
        let content = r###"
            <?xml version="1.0" encoding="UTF-8"?>
            <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
                <interface>
                    <localVars/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <textDeclaration>
                                <content>
                                    PROGRAM program_0
                                    VAR
                                        x: BOOL := 0;
                                    END_VAR
                                </content>
                            </textDeclaration>
                        </data>
                    </addData>
                </interface>
                <body>
                    <FBD>
                        <inVariable localId="0" height="20" width="80" negated="false">
                            <position x="160" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>x</expression>
                        </inVariable>
                        <label localId="1" height="20" width="80" label="lbl" executionOrderId="0">
                            <position x="570" y="50"/>
                        </label>
                        <jump localId="2" height="20" width="80" label="lbl" executionOrderId="1">
                            <position x="320" y="110"/>
                            <addData>
                                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                                    <negated value="false"/>
                                </data>
                            </addData>
                        </jump>
                        <outVariable localId="3" height="20" width="80" executionOrderId="2" negated="false" storage="none">
                            <position x="790" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="4"/>
                            </connectionPointIn>
                            <expression>x</expression>
                        </outVariable>
                        <inVariable localId="4" height="20" width="80" negated="false">
                            <position x="640" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>FALSE</expression>
                        </inVariable>
                    </FBD>
                </body>
            </pou>
        "###.to_string();

        let (ast, diagnostics) =
            xml_parser::parse(&content.into(), ast::ast::LinkageType::Internal, IdProvider::default());
        assert_debug_snapshot!(ast);
        assert_debug_snapshot!(diagnostics);
    }

    #[test]
    fn negated_jump_ast() {
        let content = r###"
            <?xml version="1.0" encoding="UTF-8"?>
            <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
                <interface>
                    <localVars/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <textDeclaration>
                                <content>
                                    PROGRAM program_0
                                    VAR
                                        x: BOOL := 0;
                                    END_VAR
                                </content>
                            </textDeclaration>
                        </data>
                    </addData>
                </interface>
                <body>
                    <FBD>
                        <inVariable localId="0" height="20" width="80" negated="false">
                            <position x="160" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>x</expression>
                        </inVariable>
                        <label localId="1" height="20" width="80" label="lbl" executionOrderId="0">
                            <position x="570" y="50"/>
                        </label>
                        <jump localId="2" height="20" width="80" label="lbl" executionOrderId="1">
                            <position x="320" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="0"/>
                            </connectionPointIn>
                            <addData>
                                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                                    <negated value="true"/>
                                </data>
                            </addData>
                        </jump>
                        <outVariable localId="3" height="20" width="80" executionOrderId="2" negated="false" storage="none">
                            <position x="790" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="4"/>
                            </connectionPointIn>
                            <expression>x</expression>
                        </outVariable>
                        <inVariable localId="4" height="20" width="80" negated="false">
                            <position x="640" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>FALSE</expression>
                        </inVariable>
                    </FBD>
                </body>
            </pou>
        "###.to_string();

        let (ast, diagnostics) =
            xml_parser::parse(&content.into(), ast::ast::LinkageType::Internal, IdProvider::default());
        assert_debug_snapshot!(ast);
        assert!(diagnostics.is_empty());
    }

    #[test]
    fn unnamed_controls() {
        let content = r###"
            <?xml version="1.0" encoding="UTF-8"?>
            <pou xmlns="http://www.plcopen.org/xml/tc6_0201" name="program_0" pouType="program">
                <interface>
                    <localVars/>
                    <addData>
                        <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                            <textDeclaration>
                                <content>
                                    PROGRAM program_0
                                </content>
                            </textDeclaration>
                        </data>
                    </addData>
                </interface>
                <body>
                    <FBD>
                        <inVariable localId="0" height="20" width="80" negated="false">
                            <position x="160" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>x</expression>
                        </inVariable>
                        <label localId="1" height="20" width="80" label="" executionOrderId="0">
                            <position x="570" y="50"/>
                        </label>
                        <jump localId="2" height="20" width="80" label="" executionOrderId="1">
                            <position x="320" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="0"/>
                            </connectionPointIn>
                            <addData>
                                <data name="www.bachmann.at/plc/plcopenxml" handleUnknown="implementation">
                                    <negated value="false"/>
                                </data>
                            </addData>
                        </jump>
                        <outVariable localId="3" height="20" width="80" executionOrderId="2" negated="false" storage="none">
                            <position x="790" y="110"/>
                            <connectionPointIn>
                                <relPosition x="0" y="10"/>
                                <connection refLocalId="4"/>
                            </connectionPointIn>
                            <expression>x</expression>
                        </outVariable>
                        <inVariable localId="4" height="20" width="80" negated="false">
                            <position x="640" y="110"/>
                            <connectionPointOut>
                                <relPosition x="80" y="10"/>
                            </connectionPointOut>
                            <expression>FALSE</expression>
                        </inVariable>
                    </FBD>
                </body>
            </pou>
        "###.to_string();

        let (ast, diagnostics) =
            xml_parser::parse(&content.into(), ast::ast::LinkageType::Internal, IdProvider::default());
        assert_debug_snapshot!(ast);
        assert_debug_snapshot!(diagnostics);
    }
}
