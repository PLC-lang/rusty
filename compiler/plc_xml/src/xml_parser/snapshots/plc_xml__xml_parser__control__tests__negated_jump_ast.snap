---
source: compiler/plc_xml/src/xml_parser/control.rs
expression: ast
---
CompilationUnit {
    global_vars: [],
    units: [
        POU {
            name: "program_0",
            variable_blocks: [
                VariableBlock {
                    variables: [
                        Variable {
                            name: "x",
                            data_type: DataTypeReference {
                                referenced_type: "BOOL",
                            },
                            initializer: Some(
                                LiteralInteger {
                                    value: 0,
                                },
                            ),
                        },
                    ],
                    variable_block_type: Local,
                },
            ],
            pou_type: Program,
            return_type: None,
        },
    ],
    implementations: [
        Implementation {
            name: "program_0",
            type_name: "program_0",
            linkage: Internal,
            pou_type: Program,
            statements: [
                LabelStatement {
                    name: "lbl",
                },
                JumpStatement {
                    condition: UnaryExpression {
                        operator: Not,
                        value: ReferenceExpr {
                            kind: Member(
                                Identifier {
                                    name: "x",
                                },
                            ),
                            base: None,
                        },
                    },
                    target: ReferenceExpr {
                        kind: Member(
                            Identifier {
                                name: "lbl",
                            },
                        ),
                        base: None,
                    },
                },
                Assignment {
                    left: ReferenceExpr {
                        kind: Member(
                            Identifier {
                                name: "x",
                            },
                        ),
                        base: None,
                    },
                    right: LiteralBool {
                        value: false,
                    },
                },
            ],
            location: SourceLocation {
                span: None,
            },
            name_location: SourceLocation {
                span: None,
            },
            overriding: false,
            generic: false,
            access: None,
        },
    ],
    user_types: [],
    file_name: "<internal>",
}
